napari/components/layerlist.py

```diff
@@ -126,8 +126,46 @@ def _update_name(self, event):
         layer = event.source
         layer.name = self._coerce_name(layer.name, layer)
 
+    def __setitem__(self, key, value):
+        # we have to repeat some logic even though it happens in the superclass
+        # because we need to ensure inplace changes such as reshuffling layers still work
+        # e.g: layers[2:5] = sorted(layers[2:5])
+        old = self._list[key]
+        if isinstance(key, slice):
+            if not isinstance(value, Iterable):
+                raise TypeError(
+                    trans._(
+                        'Can only assign an iterable to slice',
+                        deferred=True,
+                    )
+                )
+            # ensure uniqueness
+            if len(value) != len(set(value)):
+                raise ValueError(
+                    trans._("LayerList cannot contain duplicate layers.")
+                )
+            for v in value:
```

Comment:
note that this breaks some iterables. it _is_ valid to provide a consumable iterable (like a generator) to `__setitem__`... you are consuming it here, which means it won't work for the super call.  Example:

```python
def setitem(value):
    for v in value:
        ... # check something
    print(list(value))  # use it

In [31]: f((1,2,3))
[1, 2, 3]

In [32]: f((x for x in (1,2,3)))
[]
```

so you need to cast it to a list or tuple before consuming it